### PR TITLE
fix(authentication): Only verify each hash once

### LINKED ISSUE #36046: [Bug]: Password update routine slowness
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

See comments in https://github.com/nextcloud/server/pull/33898

### Steps to reproduce

1. Run unit tests on master
2. Run unit tests on stable25
3. Compare speeds

### Expected behavior

No test speed regression

### Installation method

None

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
N/A
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

Bisected by @nickvergessen

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #36046 

Regression from #33898 

While the given patch heavily improves the situation, it still performs kind of a DDoS attach on the database, as all entries get updated on each authentication, which also is on basic auth.

I might be wrong, but I think `$this->hasher->verify` is meant to be `!$this->hasher->verify`, so we update the password of all tokens that do **NOT** decrypt to the current password?

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by marcelklehr** [comment]
Yep, I'm sorry. I messed up the `!`

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 1]
```diff
@@ -448,9 +448,11 @@ public function updatePasswords(string $uid, string $password) {
 		// Update the password for all tokens
 		$tokens = $this->mapper->getTokenByUser($uid);
 		$passwordHash = $this->hashPassword($password);
+		$verifiedHashes = [];
 		foreach ($tokens as $t) {
-			$publicKey = $t->getPublicKey();
-			if ($t->getPasswordHash() === null || $this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
+			if ($t->getPasswordHash() === null || isset($verifiedHashes[$t->getPasswordHash()]) || $this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
```
Comment (marcelklehr) [inline lib/private/Authentication/Token/PublicKeyTokenProvider.php:453]: Thanks! Hasher#verify still recomputes the hash everytime the verification fails, sadly. Not sure how to cirvumvent that

Comment (nickvergessen) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php]:
```diff
diff --git a/lib/private/Authentication/Token/PublicKeyTokenProvider.php b/lib/private/Authentication/Token/PublicKeyTokenProvider.php
index adac86dd073..6cf6b8f858c 100644
--- a/lib/private/Authentication/Token/PublicKeyTokenProvider.php
+++ b/lib/private/Authentication/Token/PublicKeyTokenProvider.php
@@ -447,13 +447,17 @@ class PublicKeyTokenProvider implements IProvider {
 
                // Update the password for all tokens
                $tokens = $this->mapper->getTokenByUser($uid);
-               $passwordHash = $this->hashPassword($password);
+               $newPasswordHash = null;
                $verifiedHashes = [];
                foreach ($tokens as $t) {
                        if ($t->getPasswordHash() === null || !isset($verifiedHashes[$t->getPasswordHash()]) || !$this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
+                               if ($newPasswordHash === null) {
+                                       $newPasswordHash = $this->hashPassword($password);
+                               }
+
                                $publicKey = $t->getPublicKey();
                                $t->setPassword($this->encryptPassword($password, $publicKey));
-                               $t->setPasswordHash($passwordHash);
+                               $t->setPasswordHash($newPasswordHash);
                                $t->setPasswordInvalid(false);
                                $this->updateToken($t);
                        } else {

```
Would only hashPassword if we ever need it.

Comment (marcelklehr) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php]: :+1:

File: lib/private/Authentication/Token/PublicKeyTokenProvider.php [hunk 2]
```diff
@@ -450,13 +450,14 @@ public function updatePasswords(string $uid, string $password) {
 		$passwordHash = $this->hashPassword($password);
 		$verifiedHashes = [];
 		foreach ($tokens as $t) {
-			if ($t->getPasswordHash() === null || isset($verifiedHashes[$t->getPasswordHash()]) || $this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
-				$verifiedHashes[$t->getPasswordHash() ?: ''] = true;
+			if ($t->getPasswordHash() === null || !isset($verifiedHashes[$t->getPasswordHash()]) || !$this->hasher->verify(sha1($password) . $password, $t->getPasswordHash())) {
```
[the remark this answers is not in the document]
Comment (nickvergessen) [reply lib/private/Authentication/Token/PublicKeyTokenProvider.php:453]: Hmm, right.
Let me make the code a bit more explicit.

**Review by marcelklehr — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Comment by marcelklehr** [comment]
Customer reports slowdowns on basic auth with this applied to stable25

**Comment by nickvergessen** [comment]
See above your comment?
grafik

**Comment by marcelklehr** [comment]
Yes, that is the patch that causes slowdowns

**Pull request merged**